# Исследование надежности заемщиков

Кредитному отделу банка, для построения модели кредитного скоринга, необходимо узнать влияет ли семейное положение и количество детей у клиента банка на погашение им кредитов в срок. Также, нужно определить, есть ли другие зависимости на основе статистике по платежёспособности.
**Данные для исследования** - статистика о платёжеспособности клиентов.

**Цель исследования** - определить, влияют ли на погашение кредита в срок следующие факторы:

- Семейное положение клиента
- Количество детей у клиента

**Ход исследования:**

- Изучение полученных данных
- Предобработка данных
- Проверка гипотез и выводы
-Итоги исследования

# Обзор данных

In [1]:
# импорт библиотеки pandas
import pandas as pd

In [2]:
# открываем файл с данными и копируем датасет в переменную df
try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

In [3]:
# выводим первые 20 строк файла
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

Из таблицы видны  отрицательные  значения в столбце количества дней общего трудового стажа, разные регистры в значениях столбца с уровнем образования. Посмотрим есть ли еще явные отклонения в начале и конце таблицы. Затем посмотрим общую информацию по таблице.

In [4]:
# выводим общую информацию по таблице
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


В таблице 12 столбцов. Содержимое столбцов соответствует описанию из документации.

**Описание данных из документации:**

- children — количество детей в семье
- days_employed — общий трудовой стаж в днях
- dob_years — возраст клиента в годах
- education — уровень образования клиента
- education_id — идентификатор уровня образования
- family_status — семейное положение
- family_status_id — идентификатор семейного положения
- gender — пол клиента
- income_type — тип занятости
- debt — имел ли задолженность по возврату кредитов
- total_income — ежемесячный доход
- purpose — цель получения кредита

In [5]:
display ('Число пропусков:',data.isna().sum())# вывод инф-ии по колонкам о пропусках
display ('% пропусков:',round(data.isna().sum()/len(data)*100,2)) # процент пропущенных значений в столбцах

'Число пропусков:'

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

'% пропусков:'

children             0.0
days_employed       10.1
dob_years            0.0
education            0.0
education_id         0.0
family_status        0.0
family_status_id     0.0
gender               0.0
income_type          0.0
debt                 0.0
total_income        10.1
purpose              0.0
dtype: float64

Наблюдаются пропуски в количестве 2174 (10,% от общего количества строк) по столбцам количества дней трудового стажа и ежемесячном доходе

## Предобработка данных

### Удаление пропусков

In [6]:
#Выводим количество пропущенных значений для каждого столбца
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

В двух столбцах есть пропущенные значения. Один из них — days_employed. Пропуски в этом столбце обработаем на следующем этапе. Другой столбец с пропущенными значениями — total_income — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца income_type. Например, у человека с типом занятости сотрудник пропуск в столбце total_income должен быть заполнен медианным доходом среди всех записей с тем же типом.

In [7]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. Таким артефактом будет отрицательное количество дней трудового стажа в столбце days_employed. Для реальных данных это нормально. Обработаем значения в этом столбце: заменим все отрицательные значения положительными с помощью метода abs().

In [8]:
data['days_employed'] = data['days_employed'].abs()

Для каждого типа занятости выведем медианное значение трудового стажа `days_employed` в днях.

In [9]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

In [10]:
#Выводим перечень уникальных значений столбца children.
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

В столбце children есть два аномальных значения. Удалим строки, в которых встречаются такие аномальные значения из датафрейма data.

In [11]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

Ещё раз выводим перечень уникальных значений столбца children, чтобы убедиться, что артефакты удалены.

In [12]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

Заполним пропуски в столбце days_employed медианными значениями по каждого типа занятости income_type.

In [13]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

Убедимся, что все пропуски заполнены. Проверим себя и ещё раз выведем количество пропущенных значений для каждого столбца с помощью двух методов.

In [14]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

Заменим веще]ственный тип данных в столбце total_income на целочисленный с помощью метода astype().

In [15]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

Обработаем неявные дубликаты в столбце education. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведем их к нижнему регистру.

In [16]:
data['education'] = data['education'].str.lower()

Выведем на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалим их.

In [17]:
data.duplicated().sum()

71

In [18]:
data = data.drop_duplicates()

### Категоризация данных

На основании диапазонов, указанных ниже, создадим в датафрейме data столбец total_income_category с категориями:
- 0–30000 — 'E';
- 30001–50000 — 'D';
- 50001–200000 — 'C';
- 200001–1000000 — 'B';
- 1000001 и выше — 'A'.

Например, кредитополучателю с доходом 25000 нужно назначить категорию 'E', а клиенту, получающему 235000, — 'B'. Используем собственную функцию с именем categorize_income() и метод apply().

In [19]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [20]:
data['total_income_category'] = data['total_income'].apply(categorize_income)
data.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose total_income_category  
0               покупка жилья                     B  
1     приобретение автомобиля                     C  
2               покупка жилья                     C  
3  дополнительное образование                     B  
4             сыграть свадьбу                     C

In [21]:
# Выведем на экран перечень уникальных целей взятия кредита из столбца purpose.
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Создадим функцию, которая на основании данных из столбца purpose сформирует новый столбец purpose_category, в который войдут следующие категории:
- 'операции с автомобилем',
- 'операции с недвижимостью',
- 'проведение свадьбы',
- 'получение образования'.
Например, если в столбце purpose находится подстрока 'на покупку автомобиля', то в столбце purpose_category должна появиться строка 'операции с автомобилем'.
Используем собственную функцию с именем categorize_purpose() и метод apply(). Изучимданные в столбце purpose и определим, какие подстроки помогут  правильно определить категорию.

In [22]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [23]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)
data.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose total_income_category          purpose_category  
0               покупка жилья                     B  операции с недвижимостью  
1     приобретение автомобиля                     C    операции с автомобилем  
2               покупка жилья                     C  операции с недвижимостью  
3  дополнительное образование                     B     получение образования  
4             сыграть свадьбу                     C        проведение свадьбы

### Исследовательский анализ

Проверим, есть ли зависимость между количеством детей и возвратом кредита в срок?

In [24]:
# выведем первые 10 строк таблицы
display(data.head())

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose total_income_category          purpose_category  
0               покупка жилья                     B  операции с недвижимостью  
1     приобретение автомобиля                     C    операции с автомобилем  
2               покупка жилья                     C  операции с недвижимостью  
3  дополнительное образование                     B     получение образования  
4             сыграть свадьбу                     C        проведение свадьбы

Из таблицы выше видим, что данные по трудовому стажу в днях представлены в вещественном формате, а должны быть переведены в целочисленный.

In [25]:
# вывод общей информации о данных для проверки
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21331 entries, 0 to 21524
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   children               21331 non-null  int64  
 1   days_employed          21331 non-null  float64
 2   dob_years              21331 non-null  int64  
 3   education              21331 non-null  object 
 4   education_id           21331 non-null  int64  
 5   family_status          21331 non-null  object 
 6   family_status_id       21331 non-null  int64  
 7   gender                 21331 non-null  object 
 8   income_type            21331 non-null  object 
 9   debt                   21331 non-null  int64  
 10  total_income           21331 non-null  int64  
 11  purpose                21331 non-null  object 
 12  total_income_category  21331 non-null  object 
 13  purpose_category       21331 non-null  object 
dtypes: float64(1), int64(6), object(7)
memory usage: 2.4+ 

Преобразуем данные по трудовому стажу в тип целочисленных.

In [26]:
# с помощью метода astype() преобразуем данные к типу int
data['days_employed'] = data['days_employed'].astype(int) 
data.head()

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose total_income_category          purpose_category  
0               покупка жилья                     B  операции с недвижимостью  
1     приобретение автомобиля                     C    операции с автомобилем  
2               покупка жилья                     C  операции с недвижимостью  
3  дополнительное образование                     B     получение образования  
4             сыграть свадьбу                     C        проведение свадьбы

In [27]:
# Теперь посмотрим общие сведения по данным: вывод  общих показателей
data.describe().T

count           mean            std      min       25%  \
children          21331.0       0.474005       0.752378      0.0       0.0   
days_employed     21331.0   67148.470864  139281.136753     24.0    1024.0   
dob_years         21331.0      43.277952      12.575300      0.0      33.0   
education_id      21331.0       0.816980       0.549282      0.0       1.0   
family_status_id  21331.0       0.974497       1.421787      0.0       0.0   
debt              21331.0       0.081196       0.273143      0.0       0.0   
total_income      21331.0  165342.745769   98313.985995  20667.0  107507.0   

                       50%       75%        max  
children               0.0       1.0        5.0  
days_employed       1996.0    5337.0   401755.0  
dob_years             42.0      53.0       75.0  
education_id           1.0       1.0        4.0  
family_status_id       0.0       1.0        4.0  
debt                   0.0       0.0        1.0  
total_income      142594.0  195842.0  2265604.0

**Вывод:**

**Столбец children**

Отрицательных значений нет. Минимальное количество детей - 0. Максимальное - 5.

**Столбец days_employed**

Отрицательных значений нет. 

Минимальное значение - 24 дня. Максимальное значение 401755.Это значение указано точно не в днях,так как стаж был бы равен 1100 лет. Проверяем вероятность того, что это значение в часах-в переводе в годы получаем стаж 45,86 лет. Это значение может быть корректным, так как максимальный возраст клиента 75 лет.

Приведем все аномальные значения в столбце к значениям в днях. Т.к. миниальный возраст для приема на работу с официальным трудоустройством - 16 лет, а максимальное значение возраста - 75 лет. 

тогда верхняя граница стажа получается 59 лет (75 лет-16 лет). В днях это 59*365= 21535. Все значения выше 21 535 будут ошибками. 

**Столбец dob_years**

Отрицательных значений нет.
Минимальное значение возраста в годах 0 - т.е. клиенту нет года. Это похоже на аномалию. Это некорректные значения, скорее всего таким образом были заполнены пропуски данных.
Максимальный возраст клиента- 75 лет.


**Столбец education_id**
Отрицательных значений нет. Минимальное значение-0. Максимальное значение-4.

**Столбец family_status_id**
Отрицательных значений нет. Минимальное значение-0. Максимальное значение-4.

**Столбец debt**
Отрицательных значений нет. Минимальное значение-0. Максимальное значение-1. Так как это булевы значения говорящие о том есть задолженностьпо долговым обязательствам или нет.

**Столбец total_income**

Максимальное значение достаточно большое, отрицательных значений нет.

Необходимо произвести следующие изменения с данными для дальнейшей работы:
1) привести аномальное значение в столбце с трудовым стажем в дни
2) решить вопрос, что делать с нулевыми занчениями в столбце возраст клиента, выясним природу данного значения

**Приводим аномальное значение в столбце с трудовым стажем в дни и выведем таблицы с общими данными для проверки:**

In [28]:
data.loc[data['days_employed'] > 21535, 'days_employed'] = data.loc[data['days_employed'] > 21535, 'days_employed'] / 24
display(data.describe().T)
display(data.head())

count           mean           std      min       25%  \
children          21331.0       0.474005      0.752378      0.0       0.0   
days_employed     21331.0    4596.602298   5350.914876     24.0    1024.0   
dob_years         21331.0      43.277952     12.575300      0.0      33.0   
education_id      21331.0       0.816980      0.549282      0.0       1.0   
family_status_id  21331.0       0.974497      1.421787      0.0       0.0   
debt              21331.0       0.081196      0.273143      0.0       0.0   
total_income      21331.0  165342.745769  98313.985995  20667.0  107507.0   

                       50%       75%        max  
children               0.0       1.0        5.0  
days_employed       1996.0    5337.0    18388.0  
dob_years             42.0      53.0       75.0  
education_id           1.0       1.0        4.0  
family_status_id       0.0       1.0        4.0  
debt                   0.0       0.0        1.0  
total_income      142594.0  195842.0  2265604.0

children  days_employed  dob_years education  education_id  \
0         1        8437.00         42    высшее             0   
1         1        4024.00         36   среднее             1   
2         0        5623.00         33   среднее             1   
3         3        4124.00         32   среднее             1   
4         0       14177.75         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose total_income_category          purpose_category  
0               покупка жилья                     B  операции с недвижимостью  
1     приобретение автомобиля                     C    операции с автомобилем  
2               покупка жилья                     C  операции с недвижимостью  
3  дополнительное образование                     B     получение образования  
4             сыграть свадьбу                     C        проведение свадьбы

Из таблицы видно, что максимальное значение в столбце days_employed стало 18388 дней,а это чуть больше 50 лет. Из-за преобразований в столбце получились опять вещественные значения. Переведем их к целочисленным и выведем на экран 5 первых строк таблицы.

In [29]:
data['days_employed']=data['days_employed'].astype(int)
display(data.head())

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0          14177         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose total_income_category          purpose_category  
0               покупка жилья                     B  операции с недвижимостью  
1     приобретение автомобиля                     C    операции с автомобилем  
2               покупка жилья                     C  операции с недвижимостью  
3  дополнительное образование                     B     получение образования  
4             сыграть свадьбу                     C        проведение свадьбы

**Выясним  природу нулевого значения в столбеце dob_years**

Так как в сводной таблице минимальное значение возрата клиента в годах - 0 лет, выясним сколько таких значений в столбце.

In [30]:
data.loc[data['dob_years']==0,'dob_years'].count()

100

Проверим нет ли иных значений в столбце, где возраст клиента был бы меньше 16 лет.

In [31]:
data.loc[(data['dob_years']<16)&(data['dob_years']!=0),'dob_years'].count()

0

Значение 0 говорит, что 100 строк содежат возраст заёмщика равный 0, что является ошибкой в данных. Это подтверждает предположение, что таким образом заполнялись пропуски значений.
По моему мнению заполнять данные пропуски в столбце возраста заемщиков медианными значениями для каждой группы клиентов неселесообразно, так как количество строк с пропусками составляет менее 0,5% (100 из 21331) от общего количества заемщиков.
Поэтому данные строки удалаем из датафрейма, что не повлияет на общий результат исследования.

In [32]:
data=data[data['dob_years']!=0]

In [33]:
# Проверим, остались ли нулевые значения в столбце с возрастом заемщиков:
data.loc[data['dob_years']==0,'dob_years'].count()

0

**Вывод**: Нулевые значения в колонке dob_years отсутсвуют, так как все строки с нулевыми значениями были удачно удалены.

#### **Теперь можно переходить к ответам на вопросы.**

**Есть ли зависимость между количеством детей и возвратом кредита в срок?**

Чтобы выяснить, есть ли зависимость между количеством детей и возвратом кредита в срок, необходимо вывести категории клиентов по количеству детей и построить таблицу с показателями: количество детей и наличие кредита (столбец 'debt').
Выводим количество клиентов по уникальным зачениям  столбца 'children', чтобы посмотреть разброс показателей.

In [34]:
# выводим количество клиентов по уникальным зачениям  столбца 'children'
print(data['children'].value_counts())

0    14022
1     4792
2     2039
3      328
4       41
5        9
Name: children, dtype: int64


Из данных видно, что количество клиентов без детей и с 1 ребенком в разы превышает количество клиентов, у которых 2, 3,4 и 5 детей. По этой причине объединить статистически незначимые группы с 3,4 и 5 детьми и категорию с 2-мя детьми. 

Вводим следующие категории в зависимости от количества детей:
        "нет детей"- нет детей;
        "один ребенок"- 1 ребенок;
        "многодетные"- 2 и более детей;

In [35]:
# напишем функцию, которая будет выводить категорию людей в зависимости от  количества детей:
def child_category(children):
    if children==0:
        return 'нет детей'
    if children==1:
        return 'один ребенок'
    if children>=2:
        return 'многодетные'

In [36]:
# внесем категории в новый столбец children_rank
data['children_rank']=data['children'].apply(child_category)

In [37]:
# выведем результат
display(data.head(10))

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0          14177         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   среднее             1   
8         2           6929         35    высшее             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose total_income_category          purpose_category  \
0               покупка жилья                     B  операции с недвижимостью   
1     приобретение автомобиля                     C    операции с автомобилем   
2               покупка жилья                     C  операции с недвижимостью   
3  дополнительное образование                     B     получение образования   
4             сыграть свадьбу                     C        проведение свадьбы   
5               покупка жилья                     B  операции с недвижимостью   
6           операции с жильем                     B  операции с недвижимостью   
7                 образование                     C     получение образования   
8       на проведение свадьбы                     C        проведение свадьбы   
9     покупка жилья для семьи                     C  операции с недвижимостью   

  children_rank  
0  один ребенок  
1  один ребенок  
2     нет детей  
3   многодетные  
4     нет детей  
5     нет детей  
6     нет детей  
7     нет детей  
8   многодетные  
9     нет детей

In [38]:
# Создадим сводную таблицу, где будет видно наличие или отсутсвие долга в зависимости от 
#количества детей.

debt_children = data.pivot_table(index='children_rank', columns='debt', values='gender', aggfunc='count')
# задаем столбцы
debt_children.columns = ['no_debt', 'debt']
# Считаем долю должников
debt_children['share_of_debtors'] = debt_children['debt'] / (debt_children['debt'] + debt_children['no_debt'])
# Сортируем по столбцу доля лолжников по возрастанию
debt_children.sort_values(by='share_of_debtors')

no_debt  debt  share_of_debtors
children_rank                                 
нет детей        12964  1058          0.075453
один ребенок      4351   441          0.092028
многодетные       2192   225          0.093091

**Вывод:** Как видно из таблицы доля должников без детей составляет 7,54% от общего числа клиентов в этой категории.

Доля клиентов с просроченными долгами с 1 ребенком составляет 9,20%, доля должников с 2 и более детьми чуть выше и сотавляет 9,31%.

Можно сделать вывод, что наличие детей повышает вероятность образования долговой задолженности. Это связано с непредвиденными расходами на детей.

В настоящее время правительство всячески поддерживает семьи с несколькими детьми субсидиями и льготами. Это обьясняет небольшую разницу в доле должников с 1 ребенком и многодетными клиентами.

####  Есть ли зависимость между семейным положением и возвратом кредита в срок?
Чтобы проследить зависимость между семейным положением и возвратом кредита в срок, создадим сводную таблицу.

Сперва проверим распределение граждан по семейному положению и проверим, нет ли татистически незначимых показателей.

In [39]:

#выведем соотношение количества клиентов в разбивке по семейному положению:
print(data['family_status'].value_counts())



женат / замужем          12213
гражданский брак          4113
Не женат / не замужем     2780
в разводе                 1179
вдовец / вдова             946
Name: family_status, dtype: int64


In [40]:
# Создаем таблицу методом pivot_table()
debt_status = data.pivot_table(index='family_status', columns='debt', values='gender', aggfunc='count')
# задаем столбцы
debt_status.columns = ['no_debt', 'debt']
# Считаем долю должников
debt_status['share_of_debtors'] = debt_status['debt'] / (debt_status['debt'] + debt_status['no_debt'])
# Сортируем по столбцу доля лолжников по возрастанию
debt_status.sort_values(by='share_of_debtors')

no_debt  debt  share_of_debtors
family_status                                         
вдовец / вдова             884    62          0.065539
в разводе                 1095    84          0.071247
женат / замужем          11290   923          0.075575
гражданский брак          3730   383          0.093119
Не женат / не замужем     2508   272          0.097842

**Вывод:** 

Самый минимальный процент задолженности у категории вдовец/вдова-6,55%. Есть вероятность, что в распоряжении данной категории клиентов есть имущество или денежные средства, доставшиеся от супруга.

Далее следует группа клиентов в разводе, у данной категории процент просроченной задолженности составляет 7,12%.

У женатых людей процент просроченной задолженности составляет 7,56%.

Самый высокий процент несвоевременно возвращенной задолженности у категории не женат/не замужем-9,78%, а также у людей, живущих в гражданском браке-9,3%. Это связано с тем, что в случае непредвиденных ситуаций клиент сам отвечает по своим обязательствам и "подстраховать" или помочь с выплатой некому.

#### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
Чтобы проследить наличие зависимости между уровнем дохода и возвратом кредита в срок,возьмем следующее распределение в качестве категорий уровня дохода в:

0–30000 — 'E'; 30001–50000 — 'D'; 50001–200000 — 'C'; 200001–1000000 — 'B'; 1000000 и выше — 'A'. 

Создадим сводную таблицу наличи просроченной задолженности в зависимости от категории доходов.

In [41]:
#создаем сводную таблицу
income_status=data.pivot_table(index='total_income_category', columns='debt', values='gender', aggfunc='count')
# задаем столбцы
income_status.columns = ['no_debt', 'debt']
# Считаем долю должников
income_status['share_of_debtors']=income_status['debt']/(income_status['debt']+income_status['no_debt'])
# Сортируем по столбцу доля лолжников по возрастанию
income_status.sort_values(by='share_of_debtors')

no_debt  debt  share_of_debtors
total_income_category                                 
D                          326    21          0.060519
B                         4635   353          0.070770
A                           23     2          0.080000
C                        14503  1346          0.084926
E                           20     2          0.090909

**Вывод:** Так как выборка по категориям А и Е очень мала в сравнении с остальными категориями, мы не будем их брать в анализ.

Наименьший процент долговой просрочки у клиентов с доходом категории D(30001–50000) и составляет 6%.
Далее следует категория клиентов В с доходом 200001–1000000. Доля их просроченных долгов составляет 7,07%
Наибольшие просрочки по уплате долговых обязательств у категории С с доходом от 50000 до 200000.Процент просроченных обязательств составляет 8,5%.

Если сравнить две самые многочисленные категории В(200001–1000000)и С(50001–200000)можно увидеть, что вероятность образования задолженности выше в более многочисленной группе-С,чем больше численность группы-тем выше риск образования задолженностей. 

#### Как разные цели кредита влияют на его возврат в срок?
Создадим сводную таблицу по данным о наличии просроченных долгов и категориями целей кредита. В таблице будем использовать данные столбца purpose_category.

In [42]:
purpose_status=data.pivot_table(index='purpose_category',columns='debt', values='gender', aggfunc='count')
#задаем столбцы
purpose_status.columns=['no_debt','debt']
#Считаем долю должников
purpose_status['share_of_debtors']=purpose_status['debt']/(purpose_status['debt']+purpose_status['no_debt'])
# Сортируем по столбцу доля лолжников по возрастанию
purpose_status.sort_values(by='share_of_debtors')

no_debt  debt  share_of_debtors
purpose_category                                         
операции с недвижимостью     9927   777          0.072590
проведение свадьбы           2118   181          0.078730
получение образования        3601   369          0.092947
операции с автомобилем       3861   397          0.093236

**Вывод:** 
Заемщики, которые берут кредиты на получение образования и операции с автомобилем с большей вероятностью станут должниками.

Процент просроченных долговых обязательств по этим категориям составляет 9,29 и 9,32 соответсвенно.

Люди, которые планируют проведение свадьбы реже нарушают сроки оплат, процент просроченного кредита-7,87.

При операциях с недвижимостью самый низкий процент неуплат в срок- 7,26%.

По расчетам можно сделать вывод, что самыми проблемными выглядят автокредиты и кредиты на образование.

#### Возможные причины появления пропусков в исходных данных.
В исходных данных пропуски могли возникнуть из-за неверного ввода данных или повреждения файла.
Пропуски в рабочем стаже могут возникнуть из-за того, что клиент не помнит точный срок начала своей работы и предпочтет не заполнять данный пункт. 
Пропуски в данных об уровне дохода могут возникнуть по причине того, что люди считают эту информацию очень личной и не доверяют кредитным организациям, так как бояться утески информации.

#### Комментарий, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.
Для заполнения пропусков в имеющихся данных лучше брать медианные, а не средние значения так как последние не учитывают влияние выбросов- аномальных значений, которые будут сильно влиять на средний показатель. Медианные значения не подвержены влиянию аномалий.

### Общий вывод.
В данной работе проведено исследование надёжности заёмщиков для кредитного отдела банка в целях построения модели кредитного скоринга.


*Цель исследования* — ответить на вопросы:

- Есть ли зависимость между наличием детей и возвратом кредита в срок?
- Есть ли зависимость между семейным положением и возвратом кредита в срок?
- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
- Как разные цели кредита влияют на его возврат в срок?

Этапы проведенного иследования:

1) **Обзор исходных данных:**
 - проверка качества исходных данных
 
2) **Предобработка данных:**
 - устранение дубликатов данных
 - выявление пропусков данных и аномалий в значениях
 - заполенение пропусков данных о доходах и трудовом стаже
 - корректировка аномальных  значений в данных о трудовом стаже(отрицательные значения и значения, выраженные в часах)
 - устранение аномальных значений в данных о количестве детей
 - исключение 100 строк без информации о возрасте клиентов из данных
 - формирование категорий доходов клиентов
 - формирование категорий целей взятия кредитов
 - формирование категорий клиентов по количеству детей
 
3) **Ответы на вопросы исследования:**
 - создание сводных таблиц для выявления зависимости возврата кредита в срок от следующих факторов:количество детей,семейное положение,цели кредитования,уровень доходов.
 - формирование выводов и рекомендаций по итогам исследования

В ходе исследования данных статистики платежеспособности клиентов кредитного отдела банка были выявлены существенные отклонения  в исходных данных, касающиеся сведений о трудовом стаже(столбец "days_employed").

Корректировки потребовали:

* отрицательные значения в данных об общем трудовом стаже;
* пропущенные данные в столбцах с общим трудовым стажем и данных о доходах (total_income и days_employed). Пропуски составили 10%  от общего количества данных
* частично данные по трудовому стажу пенсионеров и безработных представлены  в часах, а не днях;
Рекомендации отделу сбора и занесению информации:
- внимательно заносить/переносить данные, учитывая вышеизложенные пункты
- привести к единообразию заполение в случае отсутвия информации по какому-то параметру (Например: избегать заполнение "0" данных по возрасту клиентов, так как это может привести к искажению расчетов)


По результатам исследования выявлено, что на возврат кредита в срок влияют несколько факторов:наличие детей, семейное положение,уровень дохода,назначение кредита.


В модели кредитного скорринга необходимо учесть следующие выводы:

- наличие детей повышает вероятность задолжености
- неженатые и живущие в граждансокм браке люди с большей вероятностью допустят просрочку платежа
- клиенты со средним доходом (50001–200000) чаще закрывают долговые обязателсьва с просрочкой
- те,кто берет кредиты на получение образования и операции с автомобилем чаще становятся должниками